block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -120,7 +120,92 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
+    * @return
+    */
+  private def representation(
+      latestMessagesToUse: Option[Set[BlockHash]] = None
+  ): F[BlockDagRepresentation[F]] = {
+
+    /** trims view on the DAG to get view as per provided `latest messages` */
+    def calculateView(
+        fullDag: KeyValueDagRepresentation,
+        tips: Set[BlockHash]
+    ): F[KeyValueDagRepresentation] = {
+      val fullLMs   = fullDag.latestMessagesMap
+      val activeLMs = fullDag.latestMessagesMap.filter(lm => tips.contains(lm._2))
+      val excessLMs = fullLMs.filterNot(v => activeLMs.contains(v._1))
+
+      // This only works when tips that are going to be used are subset of full latest messages
+      // It means this API won't let to reconstruct some historical view, only truncate the most recent view,
+      // having some latest messages excluded
+      assert(tips.subsetOf(fullDag.latestMessagesMap.values.toSet))
+
+      def replaceWithParents(lvl: Set[BlockMetadata]): F[Set[BlockMetadata]] =
+        lvl.toList.flatMap(_.parents).traverse(p => fullDag.lookupUnsafe(p)).map(_.toSet)
+
+      def isInValidatorView(blockHash: BlockHash, validators: Set[Validator]): F[Boolean] =
+        (false, Seq(blockHash)).tailRecM {
+          // descendant created by validator from target set found
+          case (true, _) => true.asRight[(Boolean, Seq[BlockHash])].pure[F]
+          // the whole DAG traversed, descendant created by validator from target set not found
+          case (false, Seq()) => false.asRight[(Boolean, Seq[BlockHash])].pure[F]
+          // keep traversing
+          case (false, children) =>
+            for {
+              nextChildren <- children.toList.traverse(c => fullDag.lookupUnsafe(c))
+              matchFound   = (nextChildren.exists(c => validators.contains(c.sender)))
+            } yield (matchFound, nextChildren.map(_.sender)).asLeft[Boolean]
+        }
+
+      for {
+        excessTips       <- excessLMs.values.toList.traverse(fullDag.lookupUnsafe).map(_.toSet)
+        excessValidators = excessLMs.keys.toSet
+        toRemove <- Concurrent[F]
+                     .tailRecM[(Set[BlockMetadata], Set[BlockMetadata]), Set[BlockMetadata]](
+                       (excessTips, excessTips)
+                     ) {
+                       case (currLvl, acc) =>
+                         currLvl match {
+                           case Seq() =>
+                             acc.asRight[(Set[BlockMetadata], Set[BlockMetadata])].pure[F]
+                           case v =>
+                             for {
+                               parents <- replaceWithParents(v)
+                               nextLvl <- parents.toList.filterA(
+                                           p => isInValidatorView(p.blockHash, excessValidators).not
+                                         )
+                             } yield (nextLvl.toSet, acc ++ v).asLeft[Set[BlockMetadata]]
+                         }
+                     }
```

Comment:
It would be nice to Streamify this and expand in parallel. :wink:

Reply:
I think this is it?
https://github.com/rchain/rchain/blob/6f66e3c28d12af75eb5ebd22d7b5992f35f0753d/shared/src/main/scala/coop/rchain/shared/DagOps.scala#L9-L24
I guess I made it too complex